text/2020-03-07-copr-chunk.md

```diff
@@ -0,0 +1,214 @@
+
+```rust
+#[derive(Debug)]
+pub enum RpnStackNodeVectorValue<'a> {
+    Generated {
+        // TODO: Maybe box it can be faster.
+        physical_value: VectorValue,
+    },
+    Ref {
+        physical_value: &'a VectorValue,
+        logical_rows: &'a [usize],
+    },
+}
+```
+
+Currently, the definition of `VectorValue` is:
+
+```rust
+#[derive(Debug, PartialEq, Clone)]
+pub enum VectorValue {
+    Int(Vec<Option<Int>>),
+    Real(Vec<Option<Real>>),
+    Decimal(Vec<Option<Decimal>>),
+    // TODO: We need to improve its performance, i.e. store strings in adjacent memory places
+    Bytes(Vec<Option<Bytes>>),
+    DateTime(Vec<Option<DateTime>>),
+    Duration(Vec<Option<Duration>>),
+    Json(Vec<Option<Json>>),
+}
+```
+
+We should modify the `VectorValue` to:
+
+```rust
+#[derive(Debug, PartialEq, Clone)]
+pub enum VectorValue {
+    Int(ChunkedVec<Int>),
+    Real(ChunkedVec<Real>),
+    Decimal(ChunkedVec<Decimal>),
+    Bytes(ChunkedVec<Bytes>),
+    DateTime(ChunkedVec<DateTime>),
+    Duration(ChunkedVec<Duration>),
+    Json(ChunkedVec<Json>),
+}
+```
+
+`ChunkedVec` implemented [chunk column format](https://pingcap.com/blog-cn/tidb-source-code-reading-10/).
+
+The original definition of `Bytes` is `type Bytes = Vec<u8>`, we have to introduce `type BytesRef<'a> = &'a [u8]` as type of input arguments due to the memory of `Bytes` is actually owned by `ChunkedVec`.
+
+The `ScalarRef` is impossible to keep compatibility.
+
+```rust
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum ScalarValueRef<'a> {
+    // Fixed sized.
+    Int(&'a Option<Int>),
+    // ...
+    Bytes(Option<&Bytes>),
+    Json(Option<&Json>),
+}
+```
+
+We have to modify it to:
+
+```rust
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum ScalarValueRef<'a> {
+    // Fixed sized.
+    Int(Option<&'a Int>),
+    // ...
+    Bytes(Option<BytesRef>),
+    Json(Option<JsonRef>),
+}
+```
+
+We also have to modify all rpn_function signatures.
+
+For fixed size type:
+
+```rust
+// From
+#[rpn_fn]
+pub fn logical_and(lhs: &Option<i64>, rhs: &Option<i64>) -> Result<Option<i64>>;
+
+// To
+#[rpn_fn]
+pub fn logical_and(lhs: &i64, rhs: &i64) -> Result<Option<i64>>;
+
+// Or
+// NULL will be merged to output by default. specify nullable to handle NULL manually.
+#[rpn_fn(nullable = true)]
+pub fn logical_and(lhs: Option<&i64>, rhs: Option<&i64>) -> Result<Option<i64>>;
+```
+
+For `JsonRef` and `Bytes`, we introduce some helper struct:
+
+```rust
+pub struct Guard {
+    _inner: (),
+}
+
+pub trait Writer<T> {
+    write(self, Option<T>) -> Guard;
+}
+
+pub struct PartialGuard {
+    w: PartialWriter,
+}
+
+impl PartialGuard {
+    partial_write(&self, T);
+    finish(self) -> Guard;
+}
+
+pub trait PartialWriter<T>: Writer<T> {
+    begin(self) -> PartialGuard;
+}
+```
+
+The definition of `Writer` and `PartialWriter` ensures some properties:
+
+1. `Guard` is zero-overhead, and the only way to achieve a `Guard` is calling `write` or `finish`.
```

Comment:
Need to mention that `Guard` ensures `write` or `finish` has been called at least once in the implementation of the coprocessor. I think the name is confusing. `Guard` does not guard any resources. It is proof that `write` or `finish` has been called. Maybe we can call it `evidence`/`witness`?

Reply:
@breeswish PTAL